AFS initiating survey dicts changes

aspina7 opened the pull request:
just starting out on changing the survey dictionaries from Dharma to Kobo. We are still waiting on the dictionary for nutrition.

aspina7 commented on R/msf_dict_survey.R:
```diff
@@ -21,76 +20,84 @@ msf_dict_survey <- function(disease, name = "MSF-survey-dict.xlsx",
+    if (utils::packageVersion("tidyr") > "0.8.99") {
+      squished <- tidyr::nest(squished, options = dplyr::starts_with("choices_"))
+    } else {
+      ## not sure this is going to be the case anymore?? can probably delete
```
should we just set dependence on tidyr > 0.8.99 ?? and remove this else statement

zkamvar replied:
I don't necessarily want to risk breaking this on older MSF computers.

aspina7 commented on R/msf_dict_survey.R:
```diff
@@ -21,76 +20,84 @@ msf_dict_survey <- function(disease, name = "MSF-survey-dict.xlsx",
+  # dat_dict <- dplyr::mutate(dat_dict, option_order_in_set = seq(dplyr::n()))
 
-  dat_dict <- dplyr::group_by(dat_dict, !!quote(column_name))
+  ## recode type to fit to DHIS2 (I dont think this is necessary?)
```
dont think this is necessary - will circle back and delete once get through all of dict adding. Just a note to self incase forget

aspina7 commented:
This PR is necessary in order for the survey template updates in Sitrep to function.

zkamvar commented on DESCRIPTION:
```diff
@@ -30,7 +30,7 @@ URL: https://r4epis.netlify.com, https://github.com/R4EPI/epidict, https://r4epi
 Imports: 
     tibble,
-    readxl,
+    rio,
```
No. Using {rio} imports a whole buttload of packages that we do not use. The reason why we didn't use {rio} initially is because it gives different outputs (tibble, data.frame, data.table) depending on the input. At least with {readxl} we know that the output will be a tibble.

zkamvar commented on R/gen_msf_data.R:
```diff
@@ -117,98 +130,185 @@ gen_msf_data <- function(dictionary, dat_dict, is_survey, varnames = "data_eleme
+    dis_output$age_months <- dis_output$age_months_calc
+
+    # calculated months based of year
+    dis_output$age_months_calc[
```
try not to stack so many evaluations into one command. It makes things more difficult to read. Instead, create new variables.

zkamvar commented on R/gen_data.R:
```diff
@@ -8,16 +8,16 @@
 #' @param dictionary Specify which dictionary you would like to use.
 #'
 #' @param varnames Specify name of column that contains variable names.
-#'   If `dictionary` is a survey, `varnames` needs to be "column_name"`.
```
The tests need to be updated for this to work.

zkamvar commented on R/gen_msf_data.R:
```diff
@@ -242,22 +342,94 @@ gen_msf_data <- function(dictionary, dat_dict, is_survey, varnames = "data_eleme
+
+
+    # add in age if routine vaccinated
+    rout_vaccinated <- dis_output$routine_vacc %in% c("verbal", "card")
```
Everything below here is repetitive and should be turned into a function.

zkamvar commented on R/utils-generators.R:
```diff
@@ -47,10 +48,10 @@ gen_ages <- function(dis_output, numcases, set_age_na = TRUE) {
   if (has_value(age_month_var) && length(U2_YEARS) > 0 && sum(U2_YEARS) > 0) {
     # age_month
-    months <- sample_age(24L, length(U2_YEARS))
+    months <- sample_age(year_cutoff + 1 * 12, length(U2_YEARS))
```
Remember that R respects order of operations
```suggestion
    months <- sample_age((year_cutoff + 1) * 12, length(U2_YEARS))
```

zkamvar commented on R/utils-generators.R:
```diff
@@ -65,7 +66,8 @@ gen_ages <- function(dis_output, numcases, set_age_na = TRUE) {
   if (has_value(age_day_var) && length(U2_MONTHS) > 0 && sum(U2_MONTHS) > 0) {
     # age_days
-    dis_output[[age_day_var]][U2_MONTHS] <- sample_age(60L, length(U2_MONTHS))
+    dis_output[[age_day_var]][U2_MONTHS] <- sample_age(month_cutoff + 1 * 20,
```
```suggestion
    dis_output[[age_day_var]][U2_MONTHS] <- sample_age((month_cutoff + 1) * 20,
```

zkamvar reviewed (changes requested):
There is a bit of work to be done as the tests don't pass (likely due to the default parameter change) and there's a bit of repetition that I'd like to smooth out. 

Additionally, I don't see a need to move to {rio} over {readxl} since these are internal functions. {rio} is nice for R users on the front end, but not so nice for us on the back end.

zkamvar commented:
I think I now understand why the tests don't pass:

The Nutrition dictionary has not yet been migrated to the Kobo version.

aspina7 commented:
> I think I now understand why the tests don't pass:
> 
> The Nutrition dictionary has not yet been migrated to the Kobo version. If I exclude Nutrition from the tests, then the tests pass.
> 
> @aspina7, can you add the updated dictionary and the associated options if you have it?

Yep nutrition survey not available yet - still waiting on it from the mhealth team. So we just need to turn of nutrition for now

zkamvar commented on R/template_data_frame.R:
```diff
@@ -25,7 +28,7 @@ template_data_frame_categories <- function(dat_dict, numcases, varnames, survey
     dis_output[[i]] <- sample(vals, numcases, replace = TRUE)
   }
 
-  multivars <- dat_dict[[varnames]][dat_dict$data_element_valuetype == "MULTI"]
```
This throws a warning with the outbreak dictionaries as they have not been moved to the Kobo system AFAICT

zkamvar commented:
I've added CI for GitHub so that we don't have to rely on the travesty that is Travis-CI and Appveyor

zkamvar commented:
This PR should not be merged yet because the Nutrition dictionary is not completed and will break things upstream.

aspina7 commented:
> This PR should not be merged yet because the Nutrition dictionary is not completed and will break things upstream.

@pbkeating

pbkeating commented:
Feck.. will follow-up… these things are a right pain in the ass

From: Alex Spina <notifications@github.com>
Sent: 01 November 2020 07:33
To: R4EPI/epidict <epidict@noreply.github.com>
Cc: Patrick Keating <Patrick.Keating@london.msf.org>; Mention <mention@noreply.github.com>
Subject: Re: [R4EPI/epidict] AFS initiating survey dicts changes (#13)


CAUTION: This email originated from outside MSF. Do not click links, open attachments or provide information unless you recognise the sender and know that the content is safe..

________________________________

This PR should not be merged yet because the Nutrition dictionary is not completed and will break things upstream.

@pbkeating<https://github.com/pbkeating>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/R4EPI/epidict/pull/13#issuecomment-720047913>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE2NEQ2QTQ64NPF2EVIES3DSNUFKNANCNFSM4RTR5S2Q>.

aspina7 commented:
@pbkeating in the mortality dictionary - both the nutrition and the malaria modules have muac, weight and height vars (i.e. duplicated). Necessary and intentional?

pbkeating commented:
> @pbkeating in the mortality dictionary:
> 
> * [ ]  both the nutrition and the malaria modules have muac, weight and height vars (i.e. duplicated). Necessary and intentional?

This was intentional as MUAC is not a standard part of the survey but maybe the variable names don't have to change since it's unlikely you would use both the nutrition and malaria modules in the same survey

> * [x]   in some of the restritction rules for mortality (i.e. the "relevant" column) - you refered to age_months_calc, which does not exist anymore. Updated in the version on github.

Removed this throughout. Thanks for catch

> * [ ]  for measles_vaccination is for `age_months <61` (i.e. five years), but months only for those under 2 years, in the code I have used age_years < 5, but should the dictionary be updated?

Thanks again, replaced this with age_years <5
> * [ ]  for `reason_no_care` in the malaria children module - the rule refers to `health_care `variable but think it should be `care_fever`

Yes, you are right - this changed after I updated all the variable names to try to standardise them somewhat and forgot to go back
> * [ ]  `malaria_treatment` is a duplicate of `malaria_treatment_preg` - presumably can just keep the first which can then be used for the add-on module too?

> * [x]  under hsb module for death section, the rules for `source_written_hsb`, `source_date_death_hsb` and `source_written_death_hsb` all refer to `source_death` but this no longer exists and should be `source_death_hsb`. Updated on github version.

pbkeating commented:
> > @pbkeating in the mortality dictionary:
> > 
> > * [ ]   both the nutrition and the malaria modules have muac, weight and height vars (i.e. duplicated). Necessary and intentional?
> 
> This was intentional as MUAC is not a standard part of the survey but maybe the variable names don't have to change since it's unlikely you would use both the nutrition and malaria modules in the same survey
> 
> > * [x]    in some of the restritction rules for mortality (i.e. the "relevant" column) - you refered to age_months_calc, which does not exist anymore. Updated in the version on github.
> 
> Removed this throughout. Thanks for catch
> 
> > * [ ]   for measles_vaccination is for `age_months <61` (i.e. five years), but months only for those under 2 years, in the code I have used age_years < 5, but should the dictionary be updated?
> 
> Thanks again, replaced this with age_years <5
> 
> > * [ ]   for `reason_no_care` in the malaria children module - the rule refers to `health_care `variable but think it should be `care_fever`
> 
> Yes, you are right - this changed after I updated all the variable names to try to standardise them somewhat and forgot to go back
> 
> > * [ ]   `malaria_treatment` is a duplicate of `malaria_treatment_preg` - presumably can just keep the first which can then be used for the add-on module too?
> 
These two questions have different answers and questions. Although malaria_treatment_infant should be the same as malaria_treatement_preg

> > * [x]   under hsb module for death section, the rules for `source_written_hsb`, `source_date_death_hsb` and `source_written_death_hsb` all refer to `source_death` but this no longer exists and should be `source_death_hsb`. Updated on github version.

pbkeating commented:
> > > @pbkeating in the mortality dictionary:
> > > 
> > > * [ ]    both the nutrition and the malaria modules have muac, weight and height vars (i.e. duplicated). Necessary and intentional?
> > 
> > 
> > This was intentional as MUAC is not a standard part of the survey but maybe the variable names don't have to change since it's unlikely you would use both the nutrition and malaria modules in the same survey
> > > * [x]     in some of the restritction rules for mortality (i.e. the "relevant" column) - you refered to age_months_calc, which does not exist anymore. Updated in the version on github.
> > 
> > 
> > Removed this throughout. Thanks for catch
> > > * [ ]    for measles_vaccination is for `age_months <61` (i.e. five years), but months only for those under 2 years, in the code I have used age_years < 5, but should the dictionary be updated?
> > 
> > 
> > Thanks again, replaced this with age_years <5 in the excel also
> > > * [ ]    for `reason_no_care` in the malaria children module - the rule refers to `health_care `variable but think it should be `care_fever`
> > 
> > 
> > Yes, you are right - this changed after I updated all the variable names to try to standardise them somewhat and forgot to go back. I've updated it in the excel also
> > > * [ ]    `malaria_treatment` is a duplicate of `malaria_treatment_preg` - presumably can just keep the first which can then be used for the add-on module too?
> 
> These two questions have different answers and questions. Although malaria_treatment_infant should be the same as malaria_treatement_preg
> 
> > > * [x]    under hsb module for death section, the rules for `source_written_hsb`, `source_date_death_hsb` and `source_written_death_hsb` all refer to `source_death` but this no longer exists and should be `source_death_hsb`. Updated on github version.

Yes, good catch. Have updated the excel also

aspina7 commented:
> @pbkeating in the mortality dictionary:
 > 
 > * [x] both the nutrition and the malaria modules have muac, weight and height vars (i.e. duplicated). Necessary and intentional?
> > This was intentional as MUAC is not a standard part of the survey but maybe the variable names don't have to change since it's unlikely you would use both the nutrition and malaria modules in the same survey

Will leave as is 

> * [x]      in some of the restritction rules for mortality (i.e. the "relevant" column) - you refered to age_months_calc, which does not exist anymore. Updated in the version on github.
> > Removed this throughout. Thanks for catch

> * [x]     for measles_vaccination is for `age_months <61` (i.e. five years), but months only for those under 2 years, in the code I have used age_years < 5, but should the dictionary be updated?
> > Thanks again, replaced this with age_years <5 in the excel also

updated github excel to read `${age_years}<5`

> * [x]     for `reason_no_care` in the malaria children module - the rule refers to `health_care `variable but think it should be `care_fever`
> > Yes, you are right - this changed after I updated all the variable names to try to standardise them somewhat and forgot to go back. I've updated it in the excel also

updated github excel too 

> * [x]     `malaria_treatment` is a duplicate of `malaria_treatment_preg` - presumably can just keep the first which can then be used for the add-on module too?
> > These two questions have different answers and questions. Although malaria_treatment_infant should be the same as malaria_treatement_preg

okay will leave - but just to note that for `malaria_treatment` there is a restriction to only pregnant ladies. 

> * [x]     under hsb module for death section, the rules for `source_written_hsb`, `source_date_death_hsb` and `source_written_death_hsb` all refer to `source_death` but this no longer exists and should be `source_death_hsb`. Updated on github version. 
> > Yes, good catch. Have updated the excel also

aspina7 commented:
@pbkeating still on mort dict: 
- [ ] `treatment_delay` - restriction should be `${care_illness_last} = "yes"`, updated on github version

aspina7 commented:
@pbkeating @zkamvar @nsbatra 
 
 Could you have a trial to make sure data isnt a mess / review code for these please? 
 
 To trial the functions: 
 - pull this project 
 - go to the survey_dicts branch 
 - Build package (under the `build `tab next to `git` click install and restard - or just ctrl+shift+b)
 - functions - demonstrations of how to use is on the [readme ](https://github.com/R4EPI/epidict#accessing-dictionaries), but note that varnames = "name" since swiching to kobo
    - `msf_dict_survey()` is for creating the dictionaries 
    - gen_data creates fake data
